### PR TITLE
Block-fetch mini-protocol

### PR DESCRIPTION
**Pull request by agaffney**
* add node-to-node mode to test program
* implement NtN chain-sync
* implement block-fetch
* support multiple messages in same muxer segment
* block hash calculation
* common functions for decoding blocks and block headers

### CODE REVIEW

File: cmd/go-ouroboros-network/chainsync.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+	oConn                 *ouroboros.Ouroboros
+	nodeToNode            bool
+	readyForNextBlockChan chan bool
+	byronEpochBaseSlot    uint64
+	byronEpochSlot        uint64
+}
+
+var syncState chainSyncState
+
+// Intersect points (last block of previous era) for each era on testnet/mainnet
```
Comment (jwitko) [inline cmd/go-ouroboros-network/chainsync.go:24]: Do you forsee this being a hassle to keep up to date?

File: cmd/go-ouroboros-network/chainsync.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	os.Exit(1)
+	return nil
+}
+
+func chainSyncDoneHandler() error {
+	return nil
+}
+
+func blockFetchStartBatchHandler() error {
+	return nil
```
Comment (jwitko) [inline cmd/go-ouroboros-network/chainsync.go:187]: I've seen a few of these now.  Are they just placeholders for future development?

**Review by jwitko — APPROVED** [review]

File: cmd/go-ouroboros-network/chainsync.go (hunk 2 above)
Comment (agaffney) [reply cmd/go-ouroboros-network/chainsync.go:187]: The library will (purposely) blow up if you don't define a callback function and receive the associated message, but we don't really care about the message here, so the callback does nothing.

File: cmd/go-ouroboros-network/chainsync.go (hunk 1 above)
Comment (agaffney) [reply cmd/go-ouroboros-network/chainsync.go:24]: A little bit, but they are really just here for convenience when testing the library. I can easily update it for a new era (for testnet at least) when I'm adding support for the new era's block format, assuming that the new block format is already in use on the blockchain.

**Pull request merged**